### PR TITLE
update support for custom atomic descriptors in format of `.sdf` and `.npz`

### PR DESCRIPTION
**Pull request by yanfeiguan**
custom atomic descriptors can be put in through `.sdf` and `.npz`. The original `.pkl` input that stores a dataframe may cause error when a different version of Pandas was used to prepare the `.pkl`.

### CODE REVIEW

File: chemprop/features/__init__.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
     rdkit_2d_normalized_features_generator, register_features_generator
 from .featurization import atom_features, bond_features, BatchMolGraph, get_atom_fdim, get_bond_fdim, mol2graph, \
     MolGraph, onek_encoding_unk, set_extra_atom_fdim
-from .utils import load_features, save_features, load_atom_features
+from .utils import load_features, save_features, load_valid_atom_features
```
Comment (mliu49) [inline chemprop/features/__init__.py:6]: I think you need to also update the function name in the `__all__` list below.

File: chemprop/data/utils.py [hunk 2]
```diff
@@ -242,6 +233,19 @@ def get_data(path: str,
             if len(all_smiles) >= max_data_size:
                 break
 
+        atom_features = None
+        atom_descriptors = None
+        if args is not None and args.atom_descriptors is not None:
+            try:
+                descriptors = load_valid_atom_features(atom_descriptors_path, [x[0] for x in all_smiles])
+            except Exception as e:
+                print('Error: failed to load or valid custom atomic descriptors: {}'.format(e))
```
Comment (mliu49) [inline chemprop/data/utils.py:242]: If this step fails, is it useful to continue (vs. raising the error and stopping)?

Also, I'm not sure what the convention in chemprop is for using print vs logging.

File: chemprop/features/utils.py [hunk 3]
```diff
@@ -56,20 +57,64 @@ def load_features(path: str) -> np.ndarray:
...
+        from rdkit.Chem import PandasTools
+        features_df = PandasTools.LoadSDF(path).drop(['ID', 'ROMol'], axis=1).set_index('SMILES')
+
+        features_df = features_df[~features_df.index.duplicated()]
+
+        # locate atomic descriptors columns
+        features_df = features_df.iloc[:, features_df.iloc[0, :].apply(lambda x: isinstance(x, str) and ',' in x).to_list()]
+
+        features_df = features_df.reindex(smiles)
+        print(features_df)
```
Comment (mliu49) [inline chemprop/features/utils.py:99]: Should this be left here?

**Review by mliu49 — COMMENTED** [review]
I added a couple comments, though I haven't done any testing. It would be great to get this in before the release!

File: chemprop/data/utils.py (hunk 2 above)
Comment (yanfeiguan) [reply chemprop/data/utils.py:242]: Good point! It shouldn't continue. It should raise an error

File: chemprop/features/utils.py (hunk 3 above)
Comment (yanfeiguan) [reply chemprop/features/utils.py:99]: Sorry for the left over!

File: chemprop/features/utils.py [hunk 4]
```diff
@@ -56,20 +57,62 @@ def load_features(path: str) -> np.ndarray:
...
+        features_df = pd.read_pickle(path)
+        if features_df.iloc[0, 0].ndim == 1:
+            features = features_df.apply(lambda x: np.stack(x.tolist(), axis=1), axis=1).tolist()
+        elif features_df.iloc[0, 0].ndim == 2:
+            features = features_df.apply(lambda x: np.concatenate(x.tolist(), axis=1), axis=1).tolist()
+        else:
+            raise ValueError(f'Atom descriptors input {path} format not supported')
+
+    elif extension == '.sdf':
+        from rdkit.Chem import PandasTools
```
Comment (swansonk14) [inline chemprop/features/utils.py:90]: Can you move this import to the top of the file?

File: chemprop/features/utils.py [hunk 5]
```diff
@@ -56,20 +57,62 @@ def load_features(path: str) -> np.ndarray:
...
+        features_df = PandasTools.LoadSDF(path).drop(['ID', 'ROMol'], axis=1).set_index('SMILES')
+
+        features_df = features_df[~features_df.index.duplicated()]
+
+        # locate atomic descriptors columns
+        features_df = features_df.iloc[:, features_df.iloc[0, :].apply(lambda x: isinstance(x, str) and ',' in x).to_list()]
+        features_df = features_df.reindex(smiles)
+        if features_df.isnull().any().any():
```
Comment (swansonk14) [inline chemprop/features/utils.py:98]: Is there a need for `.any().any()` or can you just do a single `.any()` like below?

```suggestion
        if features_df.isnull().any():
```

File: chemprop/features/utils.py [hunk 6]
```diff
@@ -56,20 +57,62 @@ def load_features(path: str) -> np.ndarray:
...
+            raise ValueError(f'Invalid custom atomic descriptors file, Nan found in data')
```
Comment (swansonk14) [inline chemprop/features/utils.py:99]: The f-string isn't necessary.

```suggestion
            raise ValueError('Invalid custom atomic descriptors file, Nan found in data')
```

File: chemprop/features/utils.py [hunk 7]
```diff
@@ -56,20 +57,62 @@ def load_features(path: str) -> np.ndarray:
...
-        features = features_df.apply(lambda x: np.concatenate(x.tolist(), axis=1), axis=1).tolist()
-    else:
-        raise ValueError(f'Atom descriptors input {path} format not supported')
 
     return features
+
+
+
+
+
```
Comment (swansonk14) [inline chemprop/features/utils.py:118]: Can you remove the extra lines and leave just a single blank line at the bottom of the file?

File: chemprop/data/utils.py [hunk 8]
```diff
@@ -242,6 +233,19 @@ def get_data(path: str,
             if len(all_smiles) >= max_data_size:
                 break
 
+        atom_features = None
+        atom_descriptors = None
+        if args is not None and args.atom_descriptors is not None:
+            try:
+                descriptors = load_valid_atom_features(atom_descriptors_path, [x[0] for x in all_smiles])
+            except Exception as e:
+                raise ValueError('Failed to load or valid custom atomic descriptors: {}'.format(e))
```
Comment (swansonk14) [inline chemprop/data/utils.py:242]: Can you switch this to an f-string like below? I just think f-strings are a little easier to read and that way we're consistent with the rest of chemprop.

```suggestion
                raise ValueError(f'Failed to load or valid custom atomic descriptors: {e}')
```

**Review by swansonk14 — CHANGES_REQUESTED** [review]
This looks great, thank you @yanfeiguan! I just have a couple tiny comments – mostly formatting things. If you can make those changes that would be great!

**Review by swansonk14 — APPROVED** [review]

**Comment by swansonk14** [comment]
@yanfeiguan I went ahead and made the changes (all except the `.any().any()` in case that is necessary) so don't worry about it!

**Pull request merged**